packages/SwingSet/src/kernel/vatManager/vat-warehouse.js

```diff
@@ -0,0 +1,257 @@
+ * @typedef {(syscall: VatSyscallObject) => ['error', string] | ['ok', null] | ['ok', Capdata]} VatSyscallHandler
+ * @typedef {{ body: string, slots: unknown[] }} Capdata
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ * @typedef { { moduleFormat: string }} Bundle
+ */
+export function makeVatWarehouse(kernelKeeper, vatLoader, policyOptions) {
+  const { sizeHint = 20 } = policyOptions || {};
+
+  /**
+   * @typedef {{
+   *   manager: VatManager,
+   *   enablePipelining: boolean, //@@@put in DB. avoid spinning up a vat that isn't pipelined, in some cases.
+   * }} VatInfo
+   * @typedef { ReturnType<typeof import('../vatTranslator').makeVatTranslators> } VatTranslators
+   */
+  const ephemeral = {
+    /** @type {Map<string, VatInfo> } key is vatID */
+    vats: new Map(),
+  };
+
+  /** @type {Map<string, VatTranslators> } */
+  const xlate = new Map();
+  /** @param { string } vatID */
+  function findOrCreateTranslators(vatID) {
+    let translators = xlate.get(vatID);
+    if (!translators) {
+      // @@@ document pre-condition makeVatTranslators needs vatKeeper?
+      // or pass in the vatKeeper too;
+      // grep for getVatKeeper and avoid this pattern.
+      // remove ephemeral table from kernelKeeper@@@
+      translators = makeVatTranslators(vatID, kernelKeeper);
+      xlate.set(vatID, translators);
+    }
+    return translators;
+  }
+
+  /**
+   * @param {string} vatID
+   * @param {boolean} recreate
+   * @returns { Promise<VatInfo> }
+   */
+  async function ensureVatOnline(vatID, recreate) {
+    const info = ephemeral.vats.get(vatID);
+    if (info) return info;
+
+    // @@@getVatKeeper -> provideVatKeeper?
+    const vatKeeper =
+      kernelKeeper.getVatKeeper(vatID) || kernelKeeper.allocateVatKeeper(vatID);
+    const { source, options } = vatKeeper.getSourceAndOptions(); // @@@comms vat source repeated?
+
+    const translators = findOrCreateTranslators(vatID);
+
+    const createVat =
+      (!recreate && vatLoader.createVatDynamically) ||
+      (kernelKeeper.getDynamicVats().includes(vatID)
+        ? vatLoader.recreateDynamicVat
+        : vatLoader.recreateStaticVat);
+    // console.log('provide: creating from bundle', vatID);
+    const manager = await createVat(vatID, source, translators, options);
+
+    const { enablePipelining = false } = options;
+
+    // TODO: load from snapshot
+    await manager.replayTranscript();
+    const result = {
+      manager,
+      translators,
+      enablePipelining,
+    };
+    ephemeral.vats.set(vatID, result);
+    return result;
+  }
+
+  /**
+   * Bring new dynamic vat online and run its (bootstrap) code.
+   *
+   * @param {string} vatID
+   */
+  async function createDynamicVat(vatID) {
+    return ensureVatOnline(vatID, false);
+  }
+
+  /** @param { typeof console.log } logStartup */
+  async function start(logStartup) {
+    const recreate = true; // note: PANIC on failure to recreate
+
```

Comment:
Let's add a note here, this is where we'd replay multiple vats in parallel to speed up relay. The original version instantiated all the vats first, then replayed each one in sequence. This version does (instantiate+replay) of each vat as a unit, which might make it less obvious how to do the parallelization.